### DISCUSSION TITLE
Poco::Condition

### DISCUSSION THREAD

**Discussion by micheleselea** [pocoproject/poco#4053] [Q&A]
Hi @obiltschnig @aleks-f I have a question about Condition class.
Why we need an external mutex, and why is used in tryWait like that?
```
{
		ScopedUnlock<Mtx> unlock(mutex, false);
		Event event;
		{
			FastMutex::ScopedLock lock(_mutex);
			mutex.unlock();
			enqueue(event);
		}
		if (!event.tryWait(milliseconds))
		{
			FastMutex::ScopedLock lock(_mutex);
			dequeue(event);
			return false;
		}
		return true;
	}
```
I can't undertand the reason

**Comment by bas524** [comment]
You can see, that standard lib does the same, for example Microsoft STL
https://github.com/microsoft/STL/blob/main/stl/inc/condition_variable#L234

**Comment by aleks-f** [comment]
To be locked until condition is satisfied, that's how condition [works](https://en.cppreference.com/w/cpp/thread/condition_variable/wait):

> Atomically unlocks lock, blocks the current executing thread, and adds it to the list of threads waiting on *this. The thread will be
> unblocked when [notify_all()](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_all) or [notify_one()](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one)
> is executed. It may also be unblocked spuriously. When
> unblocked, regardless of the reason, lock is reacquired and wait exits.

**Comment by micheleselea** [comment]
ok understood
